packages/react-aria-components/src/Collection.tsx

```diff
@@ -519,14 +524,13 @@ export class Document<T, C extends BaseCollection<T>> extends BaseNode<T> {
     let collection = this.getMutableCollection();
     if (!collection.getItem(element.node.key)) {
       collection.addNode(element.node);
-      this.mutatedNodes.add(element);
 
       for (let child of element) {
         this.addNode(child);
       }
     }
 
-    this.queueUpdate(element);
```

Comment:
The only way the Document can be updated is by React as a result of a render. The `useLayoutEffect` runs after all DOM mutations are done by React (into the fake DOM). If there are dirty nodes at that point, we need to trigger a second render to update the real DOM. Moving it to the layout effect was necessary because this was queuing updates before the DOM was fully updated which left it in an inconsistent state.